<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bayesian Multivariate Analysis of Summary Statistics • bmass</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Bayesian Multivariate Analysis of Summary Statistics">
<meta property="og:description" content="Multivariate tool for analyzing genome-wide association
    study results in the form of univariate summary statistics. The 
    goal of 'bmass' is to comprehensively test all possible multivariate
    models given the phenotypes and datasets provided. Multivariate
    models are determined by assigning each phenotype to being either
    Unassociated (U), Directly associated (D) or Indirectly associated
    (I) with the genetic variant of interest. Test results for each model 
    are presented in the form of Bayes factors, thereby allowing direct
    comparisons between models. The underlying framework implemented
    here is based on the modeling developed in &quot;A Unified Framework 
    for Association Analysis with Multiple Related Phenotypes&quot;,
    M. Stephens (2013) &lt;doi:10.1371/journal.pone.0065245&gt;.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bmass</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/bmassIntro.1.SimulatedData.html">bmass Introduction -- Simulated Data</a>
    </li>
    <li>
      <a href="articles/bmassIntro.2.RealData.html">bmass Introduction -- Real Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mturchin20/bmass">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="bmass-bayesian-multivariate-analysis-of-summary-statistics" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bmass-bayesian-multivariate-analysis-of-summary-statistics" class="anchor"></a>bmass: Bayesian multivariate analysis of summary statistics</h1></div>

<p>The <code>bmass</code> R package provides accessible functions for running the algorithms described in <a href="https://doi.org/10.1371/journal.pone.0065245">Stephens 2013 PLoS ONE</a> and applied to multiple large, publicly available GWAS datasets in <a href="https://doi.org/10.1371/journal.pgen.1008431">Turchin and Stephens 2019</a>. <code>bmass</code> conducts a Bayesian multivariate analysis of GWAS data using univariate association summary statistics. Output inclues whether any new SNPs are found as multivariate genome-wide significant as well as posterior probabilities of each significant SNP’s assignment to different multivariate models.</p>
<p>For more details on the results of applying <code>bmass</code> to publicly available GWAS datasets, please see <a href="https://doi.org/10.1371/journal.pgen.1008431">our paper in <em>PLOS Genetics</em></a>. For more details regarding the underlying algorthims of <code>bmass</code>, please see <a href="https://doi.org/10.1371/journal.pone.0065245">Stephens 2013 PLoS ONE</a>.</p>
<p>If you find a bug, or you have a question or feedback on our work, please post an <a href="https://github.com/mturchin20/bmass/issues">issue</a>.</p>
<div id="citing-this-work" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-this-work" class="anchor"></a>Citing this work</h2>
<p>If you find the <code>bmass</code> package or any of the source code in this repository useful for your work, please cite:</p>
<blockquote>
<p>Turchin MC and Stephens M (2019) “Bayesian multivariate reanalysis of large genetic studies identifies many new associations.” <em>PLOS Genetics</em> 15(10): e1008431. <a href="https://doi.org/10.1371/journal.pgen.1008431">doi.org/10.1371/journal.pgen.1008431</a></p>
</blockquote>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>Copyright (c) 2016-2019, Michael Turchin and Matthew Stephens.</p>
<p>All source code and software in this repository are made available under the terms of the <a href="https://opensource.org/licenses/mit-license.html">MIT license</a>. See file <a href="LICENSE">LICENSE</a> for the full text of the license.</p>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h2>
<p>To install <code>bmass</code> from <a href="https://cran.r-project.org/web/packages/bmass/index.html">CRAN</a>:</p>
<pre class="{r}"><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("bmass")</a></code></pre>
<p>To install the most recent dev version of <code>bmass</code> from <a href="https://github.com/mturchin20/bmass">github</a>:</p>
<pre class="{r}"><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("devtools")
devtools::install_github("mturchin20/bmass@v1.0.3", build_vignettes=TRUE)</a></code></pre>
<p>Once you have installed the package, load the package in R:</p>
<pre class="{r}"><code><a href="https://www.rdocumentation.org/packages/base/topics/library">library("bmass")</a></code></pre>
<p>Next, view and run the example code provided in the first <a href="http://mturchin20.github.io/bmass/articles/bmassIntro.1.SimulatedData.html">introductory vignette</a> using simulated data. A second, more advanced <a href="http://mturchin20.github.io/bmass/articles/bmassIntro.2.RealData.html">introductory vignette</a> is also available that involves downloading, processing, and analyzing the <a href="http://csg.sph.umich.edu/willer/public/lipids2013/">GlobalLipids 2013</a> data.</p>
</div>
<div id="credits" class="section level2">
<h2 class="hasAnchor">
<a href="#credits" class="anchor"></a>Credits</h2>
<p>The <code>bmass</code> R package was developed by <a href="http://home.uchicago.edu/mturchin20/index.html">Michael Turchin</a> at the <a href="https://www.uchicago.edu">University of Chicago</a>, with contributions from <a href="https://pcarbo.github.io">Peter Carbonetto</a> and <a href="http://stephenslab.uchicago.edu">Matthew Stephens</a>, and based on the R code provided in <a href="https://doi.org/10.1371/journal.pone.0065245">Stephens 2013 PLoS ONE</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=bmass">https://​cloud.r-project.org/​package=bmass</a>
</li>
<li>Browse source code at <br><a href="https://github.com/mturchin20/bmass">https://​github.com/​mturchin20/​bmass</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mturchin20/bmass/issues">https://​github.com/​mturchin20/​bmass/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing bmass</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Michael Turchin <br><small class="roles"> Author, maintainer </small>  </li>
<li>Matthew Stephens <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=bmass"><img src="https://www.r-pkg.org/badges/version/bmass" alt="CRAN status badge"></a></li>
<li><a href="https://travis-ci.org/mturchin20/bmass"><img src="https://travis-ci.org/mturchin20/bmass.svg?branch=master" alt="Travis Build Status"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by Michael Turchin, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
