<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bmass Introduction – Simulated Data • bmass</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="bmass Introduction – Simulated Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bmass</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bmassIntro.1.SimulatedData.html">bmass Introduction -- Simulated Data</a>
    </li>
    <li>
      <a href="../articles/bmassIntro.2.RealData.html">bmass Introduction</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mturchin20/bmass">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>bmass Introduction – Simulated Data</h1>
                        <h4 class="author">Michael C Turchin</h4>
            
            <h4 class="date">2019-03-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mturchin20/bmass/blob/master/vignettes/bmassIntro.1.SimulatedData.Rmd"><code>vignettes/bmassIntro.1.SimulatedData.Rmd</code></a></small>
      <div class="hidden name"><code>bmassIntro.1.SimulatedData.Rmd</code></div>

    </div>

    
    
<p>This is a very simple, introductory vignette on how to run <code>bmass</code>. In this introductory example, we will use a small, simulated dataset to run <code>bmass</code> and explore some basic output. The dataset is provided with the package.</p>
<p>The purpose of this vignette is to quickly allow users to get up and running with <code>bmass</code>. For a more advanced introductory vignette, please see [][here], where we download and use the GlobalLipids 2013 dataset as an example.</p>
<div id="running-bmass" class="section level2">
<h2 class="hasAnchor">
<a href="#running-bmass" class="anchor"></a>Running <code>bmass</code>
</h2>
<p>For this introductory vignette, we will be using <code>bmass_TestData1</code>, <code>bmass_TestData2</code>, and <code>bmass_TestSigSNPs</code>. These are simulated datasets created for the purposes of unit testing and this vignette. To load up ‘bmass’ and the datasets, run the following:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"bmass"</span>);</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(bmass_TestData1, bmass_TestData2, bmass_TestSigSNPs)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">Phenotypes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"bmass_TestData1"</span>, <span class="st">"bmass_TestData2"</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">bmass_TestData1 &lt;-<span class="st"> </span>bmass_TestData1[bmass_TestData1<span class="op">$</span>pValue <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>,]</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">bmass_TestData2 &lt;-<span class="st"> </span>bmass_TestData2[bmass_TestData2<span class="op">$</span>pValue <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>,]</a></code></pre></div>
<p>First, we’ll take a look at the format of one of our simulated datasets:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(bmass_TestData1)</a></code></pre></div>
<pre><code>##   Chr    BP Marker  MAF A1 A2 Direction  pValue    N
## 1   1  1000    rs1 0.20  A  G         - 6.0e-13 2500
## 2   1  2000    rs2 0.10  G  T         - 7.6e-02 2467
## 3   2  3000    rs3 0.06  T  G         + 3.0e-09 2761
## 4   3  4000    rs4 0.40  C  A         + 5.6e-03 2310
## 5   3 15000    rs5 0.40  C  T         + 1.0e-07 2410
## 6   3 21000    rs6 0.37  G  A         + 5.0e-08 2582</code></pre>
<p>Here you should see ten columns: Chr, BP, Marker, MAF, A1, A2, Direction, pValue, N. <code>bmass</code> expects each input dataset (representing a single phenotype GWAS) to contain these 9 columns and with these specific headers. There descriptions are as follows –</p>
<p>Chr: BP: Marker: MAF: A1: A2: Direction: pValue: N:</p>
<p>Now to run <code>bmass</code>, all we use is the following code:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">bmassResults &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bmass.html">bmass</a></span>(Phenotypes, bmass_TestSigSNPs);</a></code></pre></div>
<p>and included in this output is the following:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(bmassResults)</a></code></pre></div>
<pre><code>##                       Length Class  Mode     
## MarginalSNPs            3    -none- list     
## PreviousSNPs            4    -none- list     
## NewSNPs                 3    -none- list     
## LogFile                19    -none- character
## ZScoresCorMatrix        4    -none- numeric  
## Models                 18    -none- numeric  
## ModelPriors           126    -none- numeric  
## GWASlogBFMinThreshold   1    -none- numeric</code></pre>
</div>
<div id="length-class-mode" class="section level2">
<h2 class="hasAnchor">
<a href="#length-class-mode" class="anchor"></a>Length Class Mode</h2>
</div>
<div id="marginalsnps-3--none--list" class="section level2">
<h2 class="hasAnchor">
<a href="#marginalsnps-3--none--list" class="anchor"></a>MarginalSNPs 3 -none- list</h2>
</div>
<div id="previoussnps-4--none--list" class="section level2">
<h2 class="hasAnchor">
<a href="#previoussnps-4--none--list" class="anchor"></a>PreviousSNPs 4 -none- list</h2>
</div>
<div id="newsnps-3--none--list" class="section level2">
<h2 class="hasAnchor">
<a href="#newsnps-3--none--list" class="anchor"></a>NewSNPs 3 -none- list</h2>
</div>
<div id="logfile-19--none--character" class="section level2">
<h2 class="hasAnchor">
<a href="#logfile-19--none--character" class="anchor"></a>LogFile 19 -none- character</h2>
</div>
<div id="zscorescormatrix-4--none--numeric" class="section level2">
<h2 class="hasAnchor">
<a href="#zscorescormatrix-4--none--numeric" class="anchor"></a>ZScoresCorMatrix 4 -none- numeric</h2>
</div>
<div id="models-18--none--numeric" class="section level2">
<h2 class="hasAnchor">
<a href="#models-18--none--numeric" class="anchor"></a>Models 18 -none- numeric</h2>
</div>
<div id="modelpriors-126--none--numeric" class="section level2">
<h2 class="hasAnchor">
<a href="#modelpriors-126--none--numeric" class="anchor"></a>ModelPriors 126 -none- numeric</h2>
</div>
<div id="gwaslogbfminthreshold-1--none--numeric" class="section level2">
<h2 class="hasAnchor">
<a href="#gwaslogbfminthreshold-1--none--numeric" class="anchor"></a>GWASlogBFMinThreshold 1 -none- numeric</h2>
<p>As you can see, there are multiple output variables corresponding to different components of the results. For the purposes of this vignette, we will only touch on the first three variables shown from this <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary()</a></code> output: <code>PreviousSNPs</code>, <code>MarginalSNPS</code>, and <code>NewSNPs</code>. <code>PreviousSNPs</code> refers to information on the input GWAS SNPs we used for the analysis, <code>MarginalSNPs</code> refers to information on the the marginally significant SNPs that were used in the final steps of the analysis, and <code>NewSNPs</code> refers to information on any SNPs that reach multivariate genome-wide levels of significance.</p>
<p>The main result most users will be immediately interested in is whether there are any new variants identified as genome-wide multivariate significant, and this can be determined by accessing the <code>NewSNPs</code> sublist:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(bmassResults<span class="op">$</span>NewSNPs)</a></code></pre></div>
<pre><code>##            Length Class      Mode   
## SNPs       22     data.frame list   
## logBFs     18     -none-     numeric
## Posteriors 18     -none-     numeric</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(bmassResults<span class="op">$</span>NewSNPs<span class="op">$</span>SNPs)</a></code></pre></div>
<pre><code>##  [1] ChrBP                     Chr                      
##  [3] BP                        Marker                   
##  [5] MAF                       A1                       
##  [7] bmass_TestData1_A2        bmass_TestData1_Direction
##  [9] bmass_TestData1_pValue    bmass_TestData1_N        
## [11] bmass_TestData1_ZScore    bmass_TestData2_Direction
## [13] bmass_TestData2_pValue    bmass_TestData2_N        
## [15] bmass_TestData2_ZScore    GWASannot                
## [17] mvstat                    mvstat_log10pVal         
## [19] unistat                   unistat_log10pVal        
## [21] Nmin                      logBFWeightedAvg         
## &lt;0 rows&gt; (or 0-length row.names)</code></pre>
<p>Now with our four <code>GlobalLipids2013</code> files properly formatted and QCed, we are able to run <code>bmass</code>. To do so, we will also need the list of published univariate GWAS SNPs from the <code>GlobalLipids2013</code> publication (presented as two columns per SNP, chromosome and basepair position) – these have been provided in the <code>data</code> subdirectory (<code>GlobalLipids2013.GWASsnps.txt</code>). To run <code>bmass</code> with our prepared input datafiles and our list of univariate GWAS SNPs, we run the following R code:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/base/topics/library">library("bmass");
HDL &lt;- read.table("jointGwasMc_HDL.formatted.QCed.txt.gz", header=T);
LDL &lt;- read.table("jointGwasMc_LDL.formatted.QCed.HDLMatch.txt.gz", header=T);
TG &lt;- read.table("jointGwasMc_TG.formatted.QCed.HDLMatch.txt.gz", header=T); 
TC &lt;- read.table("jointGwasMc_TC.formatted.QCed.HDLMatch.txt.gz", header=T);
load("bmassDirectory/data/GlobalLipids2013.GWASsnps.rda");
Phenotypes &lt;- c("HDL", "LDL", "TG", "TC");
bmassResults &lt;- bmass(Phenotypes, GlobalLipids2013.GWASsnps);</a></code></pre>
<p>Note that to run <code>bmass</code> we supply a vector of phenotype names that also correspond to the variable names referencing each phenotype’s respective data file. Also note that you will see a warning that one or more of the <code>GlobalLipids2013</code> datafiles contains p-values smaller than the threshold at which R can handle converting them into logarthmic scale – this is to be expected.</p>
</div>
<div id="accessing-exploring-results" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-exploring-results" class="anchor"></a>Accessing &amp; exploring results</h2>
<p><code>bmass</code> returns a list containing multiple areas of information, including separate sublists pertaining to the previous univariate GWAS SNPs used for training the multivariate model priors and the new multivariate SNPs (if any) bmass has idenfitifed as significant. Run <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary(bmassResults)</a></code> and you should see the following:</p>
<pre><code>&gt; summary(bmassResults)
                      Length Class  Mode     
MarginalSNPs             3   -none- list     
PreviousSNPs             4   -none- list     
NewSNPs                  3   -none- list     
LogFile                 20   -none- character
ZScoresCorMatrix        16   -none- numeric  
Models                 324   -none- numeric  
ModelPriors           1134   -none- numeric  
GWASlogBFMinThreshold    1   -none- numeric  </code></pre>
<p>We will touch on all the above outputs.</p>
<div id="newsnps" class="section level4">
<h4 class="hasAnchor">
<a href="#newsnps" class="anchor"></a>NewSNPs</h4>
<p>The <code>NewSNPs</code> sublist contains information pertaining to the new significant multivariate associations found by <code>bmass</code>, if any were identified. Running <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary(bmassResults$NewSNPs)</a></code> you should see the following:</p>
<pre><code>&gt; summary(bmassResults$NewSNPs)
           Length Class      Mode   
SNPs         30   data.frame list   
logBFs     5427   -none-     numeric
Posteriors 5427   -none-     numeric</code></pre>
<p><code>bmassResults$NewSNPs$SNPs</code> contains the ‘main’ <code>bmass</code> output of interest – the list of new multivariate associations found by <code>bmass</code> and those SNPs’ related information. The format of this output appears as follows:</p>
<pre><code>&gt; head(bmassResults$NewSNPs$SNPs, n=3)
              ChrBP Chr        BP    Marker    MAF A1 HDL_A2 HDL_Direction
1704   10_101902054  10 101902054 rs2862954 0.4631  C      T             +
72106    10_5839619  10   5839619 rs2275774 0.1781  G      A             +
118903 11_109521729  11 109521729  rs661171 0.2876  T      G             +
       HDL_pValue  HDL_N HDL_ZScore LDL_Direction LDL_pValue    LDL_N
1704    1.287e-06 186893   4.841751             +  5.875e-01 172821.0
72106   7.601e-07 179144   4.945343             -  7.773e-05 165198.0
118903  1.705e-06 186946   4.785573             +  1.653e-02 172877.9
       LDL_ZScore TG_Direction TG_pValue     TG_N TG_ZScore TC_Direction
1704    0.5424624            +  0.013930 177587.1  2.459063            +
72106  -3.9512933            -  0.001035 169853.0 -3.280836            -
118903  2.3969983            -  0.155800 177645.0 -1.419340            +
       TC_pValue   TC_N TC_ZScore GWASannot   mvstat mvstat_log10pVal  unistat
1704   2.526e-04 187083  3.659609         0 48.70946         9.173067 23.44255
72106  1.911e-02 179333 -2.343378         0 38.26288         7.004804 24.45642
118903 1.785e-05 187131  4.290215         0 37.84098         6.917765 22.90171
       unistat_log10pVal     Nmin logBFWeightedAvg
1704            5.890421 172821.0         7.068306
72106           6.119129 165198.0         5.447201
118903          5.768276 172877.9         5.438490</code></pre>
<p><code>NewSNPs$logBFs</code> contains the log10 Bayes Factor (BF) for each multivariate model tested (in the form of a Model x SNP matrix, with the first columns of the matrix describing each model). <code>NewSNPs$Posteriors</code> is a similar matrix, just with each matrix entry representing the posterior probability for each model/SNP combination instead of the log10 BF.</p>
<pre><code>&gt; dim(bmassResults$NewSNPs$logBFs)
[1] 81 67
&gt; bmassResults$NewSNPs$logBFs[1:5,1:10]
     HDL LDL TG TC 10_101902054   10_5839619 11_109521729   11_13313759
[1,]   0   0  0  0     0.000000    0.0000000    0.0000000    0.00000000
[2,]   1   0  0  0  -233.831047 -235.0781367 -234.6670299 -234.15472067
[3,]   2   0  0  0     0.000000    0.0000000    0.0000000    0.00000000
[4,]   0   1  0  0     0.165855    0.3172489   -0.1100418    0.06393596
[5,]   1   1  0  0   -64.774919  -66.2959645  -69.2388829  -69.93309528
       11_45696596   11_47251202
[1,]    0.00000000    0.00000000
[2,] -231.68478695 -219.10321932
[3,]    0.00000000    0.00000000
[4,]   -0.04838241    0.04997886
[5,]  -65.59917325  -45.04269241</code></pre>
</div>
<div id="previoussnps" class="section level4">
<h4 class="hasAnchor">
<a href="#previoussnps" class="anchor"></a>PreviousSNPs</h4>
<p><code>PreviousSNPs</code> contains similar information as <code>NewSNPs</code> does, but pertaining to the set of previous univariate GWAS significant SNPs used to train the multivariate model priors and set the empirical significance threshold for weighted average BF. Running <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary(bmassResults$PreviousSNPs)</a></code> you see:</p>
<pre><code>&gt; summary(bmassResults$PreviousSNPs)
             Length Class      Mode   
logBFs       12069  -none-     numeric
SNPs            30  data.frame list   
DontPassSNPs    30  data.frame list   
Posteriors   12069  -none-     numeric</code></pre>
<p><code>PreviousSNPs$SNPs</code>, <code>PreviousSNPs$logBFs</code>, and <code>PreviousSNPs$Posteriors</code> all match the descriptions as above with <code>NewSNPs</code>. However <code>PreviousSNPs$DontPassSNPs</code> refers to any GWAS SNPs which are included in the final publicaiton list but which do meet the original study’s univariate GWAS p-value threshold based on the dataset released. This can for instance occur when a final GWAS SNP list is produced from combining a discovery dataset with a replication dataset, and the former data is the only part of the study publicly released. Therefore <code>bmass</code> keeps track of these SNPs so as not to call them a completely novel multivariate SNP, but also to not use them to train the multivariate model priors since they do not technically reach univariate GWAS significance based on the dataset provided alone.</p>
<p>The format of <code>DontPassSNPs</code> is the same seen in <code>NewSNPs$SNPs</code> and <code>PreviousSNPs$SNPs</code>.</p>
</div>
<div id="marginalsnps" class="section level4">
<h4 class="hasAnchor">
<a href="#marginalsnps" class="anchor"></a>MarginalSNPs</h4>
<p><code>MarginalSNPs</code> contains the same information as <code>NewSNPs</code> and <code>PreviousSNPs</code>, but pertaining to the set of marginally significant SNPs extracted from the intermediate <code>MergedDataSources</code> dataset (a data.frame that contains all the input phenotype files combined) based on the <code>SNPMarginalUnivariateThreshold</code> and <code>SNPMarginalMultivariateThreshold</code> values (default of 1e-6 for each). <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary(bmassResults$MarginalSNPs)</a></code> returns:</p>
<pre><code>&gt; summary(bmassResults$MarginalSNPs)
           Length Class      Mode   
SNPs          30  data.frame list   
logBFs     20493  -none-     numeric
Posteriors 20493  -none-     numeric</code></pre>
<p><code>MarginalSNPs$SNPs</code>, <code>MarginalSNPs$logBFs</code>, and <code>MarginalSNPs$Posteriors</code> are all as described above.</p>
</div>
<div id="zscorescormatrix" class="section level4">
<h4 class="hasAnchor">
<a href="#zscorescormatrix" class="anchor"></a>ZScoresCorMatrix</h4>
<p>This is the phenotype correlation matrix (V_0 hat in <code>Detailed Methods (Global Lipids Analysis)</code> of Stephens 2013 PLoS ONE) derived from extracting all the ‘null’ SNPs (abs(ZScore) &lt; 2 for every phenotype) in the dataset.</p>
<pre><code>&gt; bmassResults$ZScoresCorMatrix
           HDL_ZScore LDL_ZScore  TG_ZScore TC_ZScore
HDL_ZScore  1.0000000 -0.0872789 -0.3655508 0.1523894
LDL_ZScore -0.0872789  1.0000000  0.1607208 0.8223175
TG_ZScore  -0.3655508  0.1607208  1.0000000 0.2892982
TC_ZScore   0.1523894  0.8223175  0.2892982 1.0000000</code></pre>
</div>
<div id="models" class="section level4">
<h4 class="hasAnchor">
<a href="#models" class="anchor"></a>Models</h4>
<p>This is a matrix displaying all the model combinations possible given the set of d input phenotypes and 3 model categories of directly associated (1), indirectly associated (2), and unassociated (0). The total number of models displayed should be 3^d.</p>
<pre><code>&gt; dim(bmassResults$Models)
[1] 81  4
&gt; head(bmassResults$Models)
     HDL LDL TG TC
[1,]   0   0  0  0
[2,]   1   0  0  0
[3,]   2   0  0  0
[4,]   0   1  0  0
[5,]   1   1  0  0
[6,]   2   1  0  0</code></pre>
</div>
<div id="modelpriors" class="section level4">
<h4 class="hasAnchor">
<a href="#modelpriors" class="anchor"></a>ModelPriors</h4>
<p>This is the set of trained priors determined for each multivariate model using the previous univariate GWAS SNPs and an EM algorithm (see eq. 36 in Stephens 2013). Note that for any model which does not contain at least one phenotype in the directly associated category, the prior was automatically set to 0 (aside from the global null model of all phenotypes set to unassociated).</p>
<p>Also note that ModelPriors contains a full set of model priors for each value of the hyperparameter sigma_alpha (see “Prior on sigma_alpha” in Stephens 2013); these values are stored in <code>SigmaAlphas</code>, and by default are set to <code><a href="https://www.rdocumentation.org/packages/base/topics/c">c(0.005,0.0075,0.01,0.015,0.02,0.03,0.04,0.05,0.06,0.07,0.08,0.09,0.1,0.15)</a></code> (hence 81 models * 14 sigma_alphas = 1134 priors). To see a more interpretable form of the model priors, use the <code><a href="../reference/GetModelPriorMatrix.html">GetModelPriorMatrix()</a></code> function as shown below.</p>
<pre><code>&gt; length(bmassResults$ModelPriors)
[1] 1134
&gt; bmassResults[c("ModelPriorMatrix", "LogFile")] &lt;- GetModelPriorMatrix(Phenotypes, bmassResults$Models, bmassResults$ModelPriors, bmassResults$LogFile)
&gt; head(bmassResults$ModelPriorMatrix)
  HDL LDL TG TC      Prior Cumm_Prior OrigOrder
1   1   2  1  1 0.32744537  0.3274454        44
2   1   2  2  1 0.13788501  0.4653304        53
3   1   1  1  2 0.11727440  0.5826048        68
4   1   1  1  1 0.07801825  0.6606230        41
5   1   2  1  2 0.06210658  0.7227296        71
6   2   1  2  2 0.05698876  0.7797184        78</code></pre>
<p><code>ModelPriorMatrix</code> contains the model descriptions as provided in <code>Models</code> along with the combined trained priors across each sigma_alpha value; the <code>ModelPriomatrix</code> is sorted by decreasing model prior value and also shows the cumulative distribution of the priors as well as the original order position of each model.</p>
</div>
<div id="gwaslogbfminthreshold" class="section level4">
<h4 class="hasAnchor">
<a href="#gwaslogbfminthreshold" class="anchor"></a>GWASlogBFMinThreshold</h4>
<p>This is the empirical weighted average BF threshold (BFavg) used to determine whether any new SNPs are ‘multivariate significant’. As described in Stephens 2013, each SNP has a single summary metric combining all the information across each multivariate model tested; this summary metric is the weighted average of each model’s BF, where weights are determined from the previous univariate GWAS SNPs (eq. 8 in Stephens 2013). This summary metric is also calculated for the previous unviariate GWAS SNPs themselves, and the smallest BFavg among them becomes the threshold above which new SNPs are considered multivariate significant.</p>
<pre><code>&gt; bmassResults$GWASlogBFMinThreshold
[1] 4.289906</code></pre>
<p>So every new multivariate SNP <code>bmass</code> identifies in this <code>GlobalLipids2013</code> analysis has a BFavg value &gt;= 4.289906 (and is also more than a 1Mb window away from a previous univariate GWAS SNP).</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#running-bmass">Running <code>bmass</code></a></li>
      <li><a href="#length-class-mode">Length Class Mode</a></li>
      <li><a href="#marginalsnps-3--none--list">MarginalSNPs 3 -none- list</a></li>
      <li><a href="#previoussnps-4--none--list">PreviousSNPs 4 -none- list</a></li>
      <li><a href="#newsnps-3--none--list">NewSNPs 3 -none- list</a></li>
      <li><a href="#logfile-19--none--character">LogFile 19 -none- character</a></li>
      <li><a href="#zscorescormatrix-4--none--numeric">ZScoresCorMatrix 4 -none- numeric</a></li>
      <li><a href="#models-18--none--numeric">Models 18 -none- numeric</a></li>
      <li><a href="#modelpriors-126--none--numeric">ModelPriors 126 -none- numeric</a></li>
      <li><a href="#gwaslogbfminthreshold-1--none--numeric">GWASlogBFMinThreshold 1 -none- numeric</a></li>
      <li><a href="#accessing-exploring-results">Accessing &amp; exploring results</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael Turchin, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
