<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bmass Introduction – Simulated Data • bmass</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="bmass Introduction – Simulated Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bmass</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bmassIntro.1.SimulatedData.html">bmass Introduction -- Simulated Data</a>
    </li>
    <li>
      <a href="../articles/bmassIntro.2.RealData.html">bmass Introduction</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mturchin20/bmass">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>bmass Introduction – Simulated Data</h1>
                        <h4 class="author">Michael C Turchin</h4>
            
            <h4 class="date">2019-03-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mturchin20/bmass/blob/master/vignettes/bmassIntro.1.SimulatedData.Rmd"><code>vignettes/bmassIntro.1.SimulatedData.Rmd</code></a></small>
      <div class="hidden name"><code>bmassIntro.1.SimulatedData.Rmd</code></div>

    </div>

    
    
<p>This is a very simple, introductory vignette on how to run <code>bmass</code>. In this introductory example, we will use a small, simulated dataset to run <code>bmass</code> and explore some basic output. The dataset is provided with the package.</p>
<p>The purpose of this vignette is to quickly allow users to get up and running with <code>bmass</code>. For a more advanced introductory vignette, where we download and use the GlobalLipids 2013 dataset as an example, please see <a href="http://mturchin20.github.io/bmass/articles/bmassIntro.2.RealData.html">here</a>.</p>
<div id="running-bmass" class="section level2">
<h2 class="hasAnchor">
<a href="#running-bmass" class="anchor"></a>Running <code>bmass</code>
</h2>
<p>For this introductory vignette, we will be using <code>bmass_SimulatedData1</code>, <code>bmass_SimulatedData2</code>, and <code>bmass_SimulatedSigSNPs</code>. These are simulated datasets created for the purposes of unit testing and this vignette. To load up ‘bmass’ and the datasets, run the following:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"bmass"</span>);</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(bmass_SimulatedData1, bmass_SimulatedData2, bmass_SimulatedSigSNPs)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">Phenotypes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"bmass_SimulatedData1"</span>, <span class="st">"bmass_SimulatedData2"</span>)</a></code></pre></div>
<p>First, we’ll take a look at the format of one of our simulated datasets:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(bmass_SimulatedData1)</a></code></pre></div>
<pre><code>##   Chr    BP Marker  MAF A1 A2 Direction  pValue    N
## 1   1  1000    rs1 0.20  A  G         - 6.0e-13 2500
## 2   1  2000    rs2 0.10  G  T         - 7.6e-02 2467
## 3   2  3000    rs3 0.06  T  G         + 3.0e-09 2761
## 4   3  4000    rs4 0.40  C  A         + 5.6e-03 2310
## 5   3 15000    rs5 0.40  C  T         + 1.0e-07 2410
## 6   3 21000    rs6 0.37  G  A         + 5.0e-08 2582</code></pre>
<p>Here you should see nine columns: Chr, BP, Marker, MAF, A1, A2, Direction, pValue, N. <code>bmass</code> expects each input dataset (representing a single phenotype GWAS) to contain these 9 columns and with these specific headers. Their descriptions are as follows –</p>
<ul>
<li>Chr: Chromosome</li>
<li>BP: Basepair Position</li>
<li>Marker: Variant ID information, such as a rsID # (should be a unique value for each variant)</li>
<li>MAF: Minor Allele Frequency</li>
<li>A1: Reference allele</li>
<li>A2: AlternativeaAllele</li>
<li>Direction: Direction of effect size (a column of “+” or “-”)</li>
<li>pValue: p-Value from univariate GWAS</li>
<li>N: Sample size</li>
</ul>
<p>Now to run <code>bmass</code>, all we use is the following code:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">bmassResults &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bmass.html">bmass</a></span>(Phenotypes, bmass_SimulatedSigSNPs);</a></code></pre></div>
<p>Note that at its most basic level, we only need to provide two inputs for <code><a href="../reference/bmass.html">bmass()</a></code> to properly run: a vector of strings containing the variable names of each of our univariate summary GWAS datasets and a file containing the list of previously associated univariate genome-wide significant SNPs. The former input, the vector of strings, dictates to <code><a href="../reference/bmass.html">bmass()</a></code> where to find each dataset and what to call each phenotype being analyzed (ie the name of each variable containg the GWAS summary data is expected to correspond to the phenotype represented by that data). The latter input is just a list of each previous GWAS SNP with the following two columns of information, <code>Chr</code> and <code>BP</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">bmass_SimulatedSigSNPs</a></code></pre></div>
<pre><code>##   Chr   BP
## 1   1 1000
## 2   2 3000</code></pre>
</div>
<div id="exploring-basic-results" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-basic-results" class="anchor"></a>Exploring basic results</h2>
<p>Now to begin to get a sense of what <code><a href="../reference/bmass.html">bmass()</a></code> outputs, we can run the following:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(bmassResults)</a></code></pre></div>
<pre><code>##                       Length Class  Mode     
## MarginalSNPs            3    -none- list     
## PreviousSNPs            4    -none- list     
## NewSNPs                 3    -none- list     
## LogFile                19    -none- character
## ZScoresCorMatrix        4    -none- numeric  
## Models                 18    -none- numeric  
## ModelPriors           126    -none- numeric  
## GWASlogBFMinThreshold   1    -none- numeric</code></pre>
<p>As you can see, there are multiple output variables corresponding to different components of the results. For the purposes of this vignette, we will only touch on the first three variables shown from this <code><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary()</a></code> output: <code>PreviousSNPs</code>, <code>MarginalSNPS</code>, and <code>NewSNPs</code>. <code>PreviousSNPs</code> refers to information on the input GWAS SNPs we used for the analysis, <code>MarginalSNPs</code> refers to information on the the marginally significant SNPs that were used in the final steps of the analysis, and <code>NewSNPs</code> refers to information on any SNPs that reach multivariate genome-wide levels of significance.</p>
<p>The main result most users will be immediately interested in is whether there are any new variants identified as genome-wide multivariate significant, and this can be determined by accessing the <code>NewSNPs</code> sublist:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(bmassResults<span class="op">$</span>NewSNPs)</a></code></pre></div>
<pre><code>##            Length Class      Mode   
## SNPs       22     data.frame list   
## logBFs     27     -none-     numeric
## Posteriors 27     -none-     numeric</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(bmassResults<span class="op">$</span>NewSNPs<span class="op">$</span>SNPs)</a></code></pre></div>
<pre><code>##    ChrBP Chr   BP Marker  MAF A1 bmass_SimulatedData1_A2
## 9 4_7000   4 7000    rs8 0.15  G                       C
##   bmass_SimulatedData1_Direction bmass_SimulatedData1_pValue
## 9                              +                       7e-08
##   bmass_SimulatedData1_N bmass_SimulatedData1_ZScore
## 9                   2514                    5.391171
##   bmass_SimulatedData2_Direction bmass_SimulatedData2_pValue
## 9                              -                       6e-08
##   bmass_SimulatedData2_N bmass_SimulatedData2_ZScore GWASannot   mvstat
## 9                   2514                   -5.418801         0 145.0667
##   mvstat_log10pVal  unistat unistat_log10pVal Nmin logBFWeightedAvg
## 9         31.50084 29.36341          7.221849 2514           29.107</code></pre>
<p>As you can see in this example, we have 1 new SNP that reaches multivariate genome-wide significance. This is determined by using the smallest log BF weighted average value among the previous univariate significanct GWAS SNPs (<code>PreviousSNPS</code>):</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">bmassResults<span class="op">$</span>GWASlogBFMinThreshold</a></code></pre></div>
<pre><code>## [1] 7.41173</code></pre>
<p>Any <code>MarginalSNPs</code> with <code>logBFWeightedAvg</code> greater than this value are designated as new multivariate associations (see eq. 1 in <a href="https://www.biorxiv.org/">Turchin and Stephens bioRxiv 2019</a>).</p>
<p>For further exploration and details of <code>bmass</code> output, please continue to the second introductory vignette found <a href="http://mturchin20.github.io/bmass/articles/bmassIntro.2.RealData.html">here</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#running-bmass">Running <code>bmass</code></a></li>
      <li><a href="#exploring-basic-results">Exploring basic results</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael Turchin, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
